src/services/completions.ts

```diff
@@ -658,6 +670,225 @@ function isCheckedFile(sourceFile: SourceFile, compilerOptions: CompilerOptions)
+    };
+
+    function addValue(value: string | number) {
+        switch (typeof value) {
+            case "string":
+                existingStrings.add(value);
+                break;
+            case "number":
+                existingNumbers.add(value);
+        }
+    }
+
+    function hasValue(value: string | number | PseudoBigInt): boolean {
+        switch (typeof value) {
+            case "string":
+                return existingStrings.has(value);
+            case "number":
+                return existingNumbers.has(value);
+            case "object":
+                return existingBigInts.has(pseudoBigIntToString(value));
+        }
+    }
+}
+
+function typeNodeToExpression(typeNode: TypeNode, languageVersion: ScriptTarget, quotePreference: QuotePreference): Expression | undefined {
+    switch (typeNode.kind) {
+        case SyntaxKind.TypeReference:
+            const typeName = (typeNode as TypeReferenceNode).typeName;
+            return entityNameToExpression(typeName, languageVersion, quotePreference);
+        case SyntaxKind.IndexedAccessType:
+            const objectExpression =
+                typeNodeToExpression((typeNode as IndexedAccessTypeNode).objectType, languageVersion, quotePreference);
+            const indexExpression =
+                typeNodeToExpression((typeNode as IndexedAccessTypeNode).indexType, languageVersion, quotePreference);
+            return objectExpression
+                && indexExpression
+                && factory.createElementAccessExpression(objectExpression, indexExpression);
+        case SyntaxKind.LiteralType:
+            const literal = (typeNode as LiteralTypeNode).literal;
+            switch (literal.kind) {
+                case SyntaxKind.StringLiteral:
+                    return factory.createStringLiteral(literal.text, quotePreference === QuotePreference.Single);
+                case SyntaxKind.NumericLiteral:
+                    return factory.createNumericLiteral(literal.text, (literal as NumericLiteral).numericLiteralFlags);
+            }
+            return undefined;
+        case SyntaxKind.ParenthesizedType:
+            const exp = typeNodeToExpression((typeNode as ParenthesizedTypeNode).type, languageVersion, quotePreference);
+            return exp && (isIdentifier(exp) ? exp : factory.createParenthesizedExpression(exp));
+        case SyntaxKind.TypeQuery:
+            return entityNameToExpression((typeNode as TypeQueryNode).exprName, languageVersion, quotePreference);
+        case SyntaxKind.ImportType:
+            Debug.fail(`We should not get an import type after calling 'codefix.typeToAutoImportableTypeNode'.`);
+    }
+
+    return undefined;
+}
+
+function entityNameToExpression(entityName: EntityName, languageVersion: ScriptTarget, quotePreference: QuotePreference): Expression {
+    if (isIdentifier(entityName)) {
+        return entityName;
+    }
+    const realName = unescapeLeadingUnderscores(entityName.right.escapedText);

```

Comment:
Nit: I would say "unescaped", rather than "real" (which has been overused to the point of being meaningless).